### PR TITLE
Working

### PR DESCRIPTION
**Pull request by virome**
Broken attempt to move first part of Mysql backend into MysqlLikeBackend trait

### CODE REVIEW

File: diesel/src/mysql_like/types/mod.rs [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+#[derive(Debug, Clone, Copy, Default, SqlType, QueryId)]
+pub struct Unsigned<ST>(ST);
+
+impl<MysqlLike: MysqlLikeBackend> ToSql<Unsigned<TinyInt>, MysqlLike> for u8 {
+    fn to_sql<W: Write>(&self, out: &mut Output<W, MysqlLike>) -> serialize::Result {
+        ToSql::<TinyInt, MysqlLike>::to_sql(&(*self as i8), out)
+    }
+}
+
+impl<MysqlLike: MysqlLikeBackend> FromSql<Unsigned<TinyInt>, MysqlLike> for u8 {
```
Comment (weiznich) [inline diesel/src/mysql_like/types/mod.rs:35]: There are two ways here to tell the to fix that problem that rustc does not know that the associated type for `RawValue` is `[u8]`

* Explicitly specifying the associated type in the trait bound: `MysqlLike: MysqlLikeBackend<RawValue=[u8]>`
* Us the associated type in place of `[u8]` in the function definition: `fn from_sql(bytes: Option<&MysqlLike::RawValue>) -> deserialize::Result<Self> {`

(I would probably choose the second variant)

File: diesel/src/mysql_like/backend.rs [hunk 2]
```diff
@@ -0,0 +1,26 @@
+//! The MySQL backend
+use byteorder::NativeEndian;
+
+use backend::*;
+use sql_types::TypeMetadata;
+
+pub trait MysqlLikeBackend {
```
Comment (weiznich) [inline diesel/src/mysql_like/backend.rs:7]: I would make that trait only a marker trait (== it does not have any items)

Then both, the mysql and the mariadb backend implement the corresponding traits (`Backend` and `TypedMetadata`). Additionally `MysqlLike` is implemented, similar to the `Support*` traits.

**Review by weiznich — COMMENTED** [review]
cc @sgrif

It's fine to open this as WIP PR against diesel, so we have the final discussion in one place.

File: diesel/src/mysql_like/types/mod.rs (hunk 1 above)
Comment (virome) [reply diesel/src/mysql_like/types/mod.rs:35]: Thanks for reviewing!

The first snippet there is new to me, I didn't know about associated type bounds, that's very handy.
I think I've tried the second one alone, and it gave me compiler errors about 'not being able to array access ...RawValue ' and maybe the reason it failed to compile was that the compiler couldn't prove that RawValue was [u8] for every MysqlLikeBackend.

So now I'm thinking that I need both of your suggestions, let my give it a try.

File: diesel/src/mysql_like/backend.rs (hunk 2 above)
Comment (virome) [reply diesel/src/mysql_like/backend.rs:7]: That's what I tried initially, and adding those associated types was some experimentation. I'll be glad to get rid of them after trying the suggestions in the previous comment.

I'll also make a better branch name and make the PR into the diesel repo.

**Pull request closed without merging**